// n8n-nodes-google-speech/dist/credentials/GoogleSpeechApi.credentials.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GoogleSpeechApi = void 0;
class GoogleSpeechApi {
    constructor() {
        this.name = 'googleSpeechApi';
        this.displayName = 'Google Speech API';
        this.documentationUrl = 'https://cloud.google.com/speech-to-text/docs/quickstart-client-libraries';
        this.properties = [
            {
                displayName: 'Service Account Key',
                name: 'serviceAccountKey',
                type: 'string',
                typeOptions: {
                    password: true,
                },
                default: '',
                description: 'Full Google Cloud service account key in JSON format. Create a service account and download the JSON file from GCP Console > IAM & Admin > Service Accounts.',
                required: true,
                hint: 'Paste the entire contents of your service account JSON key file here',
            },
        ];
    }
}
exports.GoogleSpeechApi = GoogleSpeechApi;
